imap_processing/ultra/l1b/badtimes.py

```diff
@@ -25,14 +32,15 @@ def calculate_badtimes(
     badtimes_dataset : xarray.Dataset
         Dataset containing the data.
     """
-    badtimes_dict = {}
-    badtimes_dict["spin_number"] = extendedspin_dataset["spin_number"]
-    badtimes_dict["energy_bin_geometric_mean"] = extendedspin_dataset[
-        "energy_bin_geometric_mean"
-    ]
+    spins = np.setdiff1d(extendedspin_dataset["spin_number"].values, cullingmask_spins)
 
-    # TODO: add more data to badtimes_dict.
+    filtered_dataset = extendedspin_dataset.sel(spin_number=spins)
 
-    badtimes_dataset = create_dataset(badtimes_dict, name, "l1b", data_version)
+    dataset_dict = {
+        **{var: filtered_dataset[var].values for var in filtered_dataset.data_vars},
+        **{coord: filtered_dataset[coord].values for coord in filtered_dataset.coords},
+    }
```

Comment:
Two other possibilities here that might simplify things.
1. Can you pass in `filtered_dataset` directly below? `create_dataset(filtered_dataset)`, datasets are mappings so should be usable in a dict-like sense.
2. You might be able to do this all in one go using `variables`: `{var: filtered_dataset[var] for var in filtered_dataset.variables}` https://docs.xarray.dev/en/stable/generated/xarray.Dataset.variables.html